pkg/pdutil/placement_client.go

```diff
@@ -0,0 +1,193 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pdutil
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+	"path"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb/store/pdtypes"
+	"github.com/pingcap/tiflow/pkg/config"
+	"github.com/pingcap/tiflow/pkg/httputil"
+	"github.com/pingcap/tiflow/pkg/retry"
+	pd "github.com/tikv/pd/client"
+	"go.uber.org/zap"
+)
+
+const (
+	updateTypeDel = iota
+	updateTypeAdd
+)
+
+const (
+	// DefaultMaxRetry is the default max number of verifing placement rules.
+	DefaultMaxRetry   = 3
+	batchUpdatePrefix = "/pd/api/v1/config/rules/batch"
+	idQueryPrefix     = "/pd/api/v1/config/rule"
+
+	// Split the default rule by `6e000000000000000000f8` to keep metadata region
+	// isolated from the normal data area. Note that the default rules also need
+	// to be modified here to avoid rule stacking.  
+	// Ref: https://docs.pingcap.com/tidb/stable/configure-placement-rules
+	addMetaJSON = `[{
+		"group_id": "pd",
+		"id": "meta", 
+		"start_key": "",
+		"end_key": "6e00000000000000f8",
+		"role": "voter",
+		"count": 3,
+		"action": "add",
+		"delete_by_id_prefix": false
+	}, 
+	{
+		"group_id": "pd",
+		"id": "default",
+		"start_key": "6e00000000000000f8",
+		"end_key": "",
+		"role": "voter",
+		"count": 3,
```

Comment:
If user set the replicas to 5, how to consist with the the user behavior? I think it's better let PD to split meta and table data by default.

Reply:
Ref: https://docs.pingcap.com/tidb/stable/configure-placement-rules#scenario-1-use-three-replicas-for-normal-tables-and-five-replicas-for-the-metadata-to-improve-cluster-disaster-tolerance
Typically, the user-configured rule should override the default rule, which means that `override = true` and `index > 0`. Besides, although it does make more sense for pd to maintain default rules, we still need to verify the rules in cdc for compatibility with older versions of the cluster.

Reply:
User can use `pd-ctl` to `config set max-replicas 5`, which changes the default pd rule from `count:3` to `count: 5`. If user upgrade cluster and with  the latest cdc, which make the replicas to 3. I think we need test it.